src/pointer-analysis/dereference.cpp

```diff
@@ -908,10 +908,27 @@ void dereferencet::build_reference_rec(
     flags |= flag_dst_scalar;
   else if (is_array_type(type))
   {
-    log_error(
-      "Can't construct rvalue reference to array type during dereference\n"
-      "(It isn't allowed by C anyway)\n");
-    abort();
+    /*
+      contract Base
+      {
+        uint[3] x;
+        constructor()
+        {
+          if(rand) x = [1, 2, 3];
+          else x = [4, 5, 6];
+          x[3]; // out_of_bound 
+        }
+      }
+    */
+    std::string sol;
+    options.get_option("sol", sol);
+    if (sol.empty())
+    {
+      log_error(
+        "Can't construct rvalue reference to array type during dereference\n"
+        "(It isn't allowed by C anyway)\n");
+      abort();
+    }
```

Comment:
This fixes nothing.

Reply:
It just avoid the abort. The result is correct.

Reply:
What is the result?

Reply:
E.g.

```solidity
contract test {
    uint[3] public t = [0, 0]; // t = [0,0,0]

    constructor() {
        (t, t) = ([1], [2]);  // t = [2,0,0]; t=[1,0,0]
    }
}

contract test1 is test {
    function test2() public view {
        assert(t[0] == 1);
        assert(t[1] == 0);
        assert(t[2] == 0);
    }
}
```
- Command: `--unwind 10 --no-unwinding-assertions`
- Result: `VERIFICATION SUCCESSFUL`

I am not sure what's going on in the middleware & backend. 
E.g.
```solidity
 int x[3] = [1, 2, 3];
 constructor(){
     x = [4, 5, 6];
 }
```
I guess the assignment will just overwrite the x's value. The corresponding symbol table is:
```
Symbol......: sol:@C@test@F@test#19
Module......: example
Base name...: test
Mode........: C++
Type........: constructor  (struct test *)
Value.......:
{
&(this->t = { 1, 2, 3 })[0];

&(this->t = { 4, 5, 6 })[0];

}
Flags.......: lvalue
Location....: file example.sol line 4
```

Reply:
Thanks for the examples, they're helpful. The reason the dereference module doesn't handle array dereferences is because there are no C semantics for it - since C doesn't have that construct: A dereference in C always refers to either a built-in type like `int` or to a non-array struct or union (member). For the C++ frontend we got around that problem in two ways:
1. array-typed statements: https://github.com/esbmc/esbmc/blob/7b1e7f2d750f8d579db145a118338a9c52b81347/src/clang-c-frontend/clang_c_adjust_code.cpp#L45-L60
2. during a constructor call: https://github.com/esbmc/esbmc/blob/7b1e7f2d750f8d579db145a118338a9c52b81347/src/clang-cpp-frontend/clang_cpp_convert.cpp#L1184-L1190

Would that be an option for you with Solidity?

Reply:
Could they support the dynamic array? Like:
```solidity
unit[] x;

constructor()
{
  x = [1];
}

function test() public
{
  x = [1, 2, 3];
  x = [4, 5];
}
```

---

I can make the fixed array to pointer type in the front end, like the dynamic array. E.g. 
```
Symbol......: sol:@C@Base@x#3
Module......: example
Base name...: x
Mode........: C++
Type........: unsigned _ExtInt(256) *
Value.......: 0
Flags.......: lvalue file_local
Location....: file example.sol line 2

Symbol......: sol:@C@Base@F@test#20
Module......: example
Base name...: test
Mode........: C++
Type........: void (struct Base *)
Value.......:
{
this->x = &{ 1, 2, 3 }[0];

this->x = &{ 4, 5 }[0];

}
Flags.......: lvalue
Location....: file example.sol line 4
```
But I don't feel like changing the code for the fixed array as they already 'work' somehow.